hive-metastore/src/main/java/org/apache/iceberg/hive/CachedClientPool.java

```diff
@@ -87,4 +106,125 @@ public <R> R run(Action<R, IMetaStoreClient, TException> action, boolean retry)
       throws TException, InterruptedException {
     return clientPool().run(action, retry);
   }
+
+  @VisibleForTesting
+  static Key toKey(List<Supplier<Object>> suppliers) {
+    return Key.of(suppliers.stream().map(Supplier::get).collect(Collectors.toList()));
+  }
+
+  @VisibleForTesting
+  static List<Supplier<Object>> extractKeySuppliers(String cacheKeys, Configuration conf) {
```

Comment:
TBH I find the code difficult to reason about, especially given the fact that the Cache key is now essentially a `List<Object>`.

I was wondering whether it would be possible to build up a String that includes all of the relevant items in string form. 
Something like `uri:<...>_ugi:<...>_username:<...>_conf:<...>` but you'd probably need to use delimiters that are unique (and also I don't know if a string representation of `UserGroupInformation` would be unique)